### PR TITLE
AZINTS-2562 Metric Filtering ARM templates

### PR DESCRIPTION
**Pull request by aishwarya-ddog**
### What does this PR do?

Update ARM templates to help update and get metric filters - ie allow users to enable and disable Microsoft resource providers from metric collection. 
metrics_config is a json which has one attribute : excluded_resource_providers which takes in a list of Microsoft resource provider name strings. 
ARM templates let the users specify excluded metric filters directly. However, we parse it into metrics_config when we make the API call. 

### RFC
https://docs.google.com/document/d/10sDJY0CWCzlAHCvV11DCkmWftTqz2qaxUvI_VYXHmcE/edit?usp=sharing 

We are enabling users to set metric filters at resource provider level in the configuration page and through the UI. Translating this capability to ARM templates

### Testing Guidelines

Checked the ARM templates created

azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json
Screenshot 2024-07-05 at 1 37 34 PM

https://github.com/DataDog/datadog-serverless-functions/pull/816#issue-2391757166
Screenshot 2024-07-05 at 1 40 24 PM

azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json
Screenshot 2024-07-05 at 2 39 44 PM

### Types of changes

- [n] Bug fix
- [y] New feature
- [n] Breaking change
- [n] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [y] This PR's description is comprehensive
- [n] This PR contains breaking changes that are documented in the description
- [y] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [y] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [y] This PR's changes are covered by the automated tests
- [n] This PR collects user input/sensitive content into Datadog
- [y] This PR passes the integration tests (ask a Datadog member to run the tests)
- [y] This PR passes the unit tests 
- [y] This PR passes the installation tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json [hunk 1]
```diff
@@ -275,12 +275,23 @@
                 "type": "Microsoft.Common.CheckBox",
                 "label": "Collect Custom Metrics",
                 "toolTip": "This option will enable custom metric collection from all Application Insights instances within the scope of this configuration. These custom metrics will appear in datadog under the namespace <b>application_insights.custom.&lt;METRIC_NAME&gt;</b>.<br> Note that <a target='_blank' href='https://learn.microsoft.com/en-us/azure/azure-monitor/app/standard-metrics/'>Standard</a> App Insights metrics are collected automatically and are available under the <b>azure.insights_components.*</b> namespace."
+              },
+              {
+                "name": "excluded_resource_providers",
+                "type": "Microsoft.Common.TextBox",
+                "label": "Excluded Resource Providers",
+                "defaultValue": "",
+                "toolTip": "Comma-separated list of Microsoft Resource Provider names to be excluded from metrics collection.",
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json]: is a customer going to know what a resource provider is/is there any concern around case sensitivity here? an example of what we expect at the very least would be good, if not a more robust solution to validate that the customer had inputted only valid resource providers

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json [hunk 2]
```diff
@@ -275,12 +275,23 @@
...
+                  "required": false
+                },
+                "visible": true
               }
             ],
             "visible": true
           },
           {
-            "name": "resourceCollection",
+            "name": "resource,Collection",
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json]: why is there a comma here now?

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json [hunk 3]
```diff
@@ -335,7 +346,8 @@
       "automute": "[steps('datadogConfig').metrics.automute]",
       "custom_metrics_enabled": "[steps('datadogConfig').metrics.custom_metrics_enabled]",
       "resource_collection_enabled": "[steps('datadogConfig').resourceCollection.resource_collection_enabled]",
-      "cspm_enabled": "[and(steps('datadogConfig').resourceCollection.cspm_enabled, steps('datadogConfig').resourceCollection.resource_collection_enabled)]"
+      "cspm_enabled": "[and(steps('datadogConfig').resourceCollection.cspm_enabled, steps('datadogConfig').resourceCollection.resource_collection_enabled)]",
+      "excluded_resource_providers": "[steps('datadogConfig').metrics.excluded_resource_providers]"
     }
   }
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json:352]: nit: can we ensure the newline at the end of the file isnt removed (both this and the next file)

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition.json [hunk 4]
```diff
@@ -251,6 +251,17 @@
                 "type": "Microsoft.Common.CheckBox",
                 "label": "Collect Custom Metrics",
                 "toolTip": "This option will enable custom metric collection from all Application Insights instances within the scope of this configuration. These custom metrics will appear in datadog under the namespace <b>application_insights.custom.&lt;METRIC_NAME&gt;</b>.<br> Note that <a target='_blank' href='https://learn.microsoft.com/en-us/azure/azure-monitor/app/standard-metrics/'>Standard</a> App Insights metrics are collected automatically and are available under the <b>azure.insights_components.*</b> namespace."
+              },
+              {
+                "name": "excluded_resource_providers",
+                "type": "Microsoft.Common.TextBox",
+                "label": "Excluded Resource Providers",
+                "defaultValue": "",
+                "toolTip": "Comma-separated list of Microsoft Resource Provider names to be excluded from metrics collection.",
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/CreateUIDefinition.json]: same comment as before about the tooltip

File: azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json [hunk 5]
```diff
@@ -100,6 +100,13 @@
                 "description": "Whether or not to use custom metrics"
             }
         },
+        "excluded_resource_providers": {
+            "type": "array",
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json]: its a string, right?

File: azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json [hunk 6]
```diff
@@ -236,6 +246,9 @@
                         "custom_metrics_enabled": {
                             "type": "bool"
                         },
+                        "excluded_resource_providers": {
+                            "type": "array"
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json]: same as above

File: azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json [hunk 7]
```diff
@@ -313,7 +330,7 @@
...
+                                "scriptContent": "Invoke-WebRequest -SkipCertificateCheck -ContentType \"application/json\" -Method Post -Uri \"https://api.$Env:datadogSite/api/v1/integration/azure\" -Headers @{\"Accept\"=\"application/json\"; \"DD-APPLICATION-KEY\"=$Env:datadogApplicationKey; \"DD-API-KEY\"=$Env:datadogApiKey} -Body (@{\"automute\"=[bool]::Parse($Env:automute); \"resource_collection_enabled\"=[bool]::Parse($Env:resource_collection_enabled); \"cspm_enabled\"=[bool]::Parse($Env:cspm_enabled); \"custom_metrics_enabled\"=[bool]::Parse($Env:custom_metrics_enabled); \"client_id\"=$Env:clientId; \"client_secret\"=$Env:clientSecret; \"tenant_name\"=$Env:tenantName; \"source\"=$Env:source; \"management_group_id\"=$Env:managementGroupId; \"metrics_config\"=@{ \"excluded_resource_providers\" = $Env:excluded_resource_providers}} | ConvertTo-Json )"
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json:333]: this is a string, so should probably look the same as other string variables

File: azure/deploy-to-azure/azure-integration/azuredeploy.json [hunk 8]
```diff
@@ -99,6 +99,13 @@
                 "description": "Whether or not to use custom metrics"
             }
         },
+        "excluded_resource_providers": {
+            "type": "array",
```
Comment (ava-silver) [inline azure/deploy-to-azure/azure-integration/azuredeploy.json]: same comments here as the other files

**Review by ava-silver — CHANGES_REQUESTED** [review]
could we also get a couple screenshots of the UI to show what it looks like?

**Comment by ava-silver** [comment]
Also want to note here that we should only be merging once this is GA/ready for all customers

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json (hunk 1 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json]: There is no concern around case sensitivity. the backend code is made to be case insensitive. It is a good idea to validate whether the users are sending accepted resource provider names. I'll include a PR in dogweb to make the validation.

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json (hunk 2 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json]: typo. corrected

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json (hunk 3 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json:352]: done

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition.json (hunk 4 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/CreateUIDefinition.json]: done

File: azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json (hunk 5 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json]: so excluded resource providers is a list of string within the json. So kept it as an array here. I understand json is also represented by a string. Dont have a strong preference either way.

File: azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json (hunk 6 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json]: please refer the comment as above

File: azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json (hunk 7 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/azuredeploy-managementgroups.json:333]: ahh. so, changing the type from array to string will make a difference here. will do

File: azure/deploy-to-azure/azure-integration/azuredeploy.json (hunk 8 above)
Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/azuredeploy.json]: Will update datatype here.

File: azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json (hunk 1 above)
Comment (ava-silver) [reply azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json]: > There is no concern around case sensitivity.

Probably worth noting for users then, and like i mentioned would probably be worth including an example.

Comment (aishwarya-ddog) [reply azure/deploy-to-azure/azure-integration/CreateUIDefinition-managementgroups.json]: included example

**Comment by aishwarya-ddog** [comment]
https://github.com/DataDog/datadog-serverless-functions/pull/816#issuecomment-2211084354 @ava-silver 
Is this a standard practice to not release ARM templates unless fully GA'd for our team? because, from the #api-platform team, it was said to be ok to release beta features on our 3p libraries as long as we note down that it is not a fully GA'd feature.

**Comment by ava-silver** [comment]
including spaces and `[]` in the list example here is confusing, are we allowing those? i was under the impression that this would be similar to the format of the existing tag filters. if so, could we make the format of these examples align with the existing ones?

**Comment by ava-silver** [comment]
closing so i stop getting pinged, we can re-open later if/when we continue this work.

**Pull request closed without merging**